File: DSA/Stack/Code/StackImplementationArray.py
# Stack Implementation using Array in Python

class Stack:
    def __init__(self, max_size=100):
        self.arr = [0] * max_size
        self.top = -1
        self.max_size = max_size
    
    def is_full(self):
        return self.top == self.max_size - 1
    
    def is_empty(self):
        return self.top == -1
    
    def push(self, value):
        if self.is_full():
            print("Stack overflow")
            return
        self.top += 1
        self.arr[self.top] = value
        print(f"Pushed: {value}")
    
    def pop(self):
        if self.is_empty():
            print("Stack underflow")
            return -1
        value = self.arr[self.top]
        self.top -= 1
        return value
    
    def peek(self):
        if self.is_empty():
            print("Stack is empty")
            return -1
        return self.arr[self.top]
    
    def display(self):
        if self.is_empty():
            print("Stack is empty")
            return
        
        print("Stack: ", end="")
        for i in range(self.top + 1):
            print(self.arr[i], end=" ")
        print()

# Test
stack = Stack()

print("=== Stack Implementation (Array) ===\n")

stack.push(10)
stack.push(20)
stack.push(30)
stack.push(40)

stack.display()

print(f"\nPeek: {stack.peek()}")

print(f"\nPopped: {stack.pop()}")
print(f"Popped: {stack.pop()}")

stack.display()

print("\nComplexity Analysis:")
print("Push: O(1)")
print("Pop: O(1)")
print("Peek: O(1)")
print("Space: O(n)")

File: DSA/Stack/Code/StackImplementationArray.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX_SIZE 100

// Stack Implementation using Array in C
typedef struct {
    int arr[MAX_SIZE];
    int top;
} Stack;

Stack* createStack() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    stack->top = -1;
    return stack;
}

bool isFull(Stack* stack) {
    return stack->top == MAX_SIZE - 1;
}

bool isEmpty(Stack* stack) {
    return stack->top == -1;
}

void push(Stack* stack, int value) {
    if (isFull(stack)) {
        printf("Stack overflow\n");
        return;
    }
    stack->arr[++stack->top] = value;
    printf("Pushed: %d\n", value);
}

int pop(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack underflow\n");
        return -1;
    }
    return stack->arr[stack->top--];
}

int peek(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty\n");
        return -1;
    }
    return stack->arr[stack->top];
}

void display(Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty\n");
        return;
    }
    
    printf("Stack: ");
    for (int i = 0; i <= stack->top; i++) {
        printf("%d ", stack->arr[i]);
    }
    printf("\n");
}

int main() {
    Stack* stack = createStack();
    
    printf("=== Stack Implementation (Array) ===\n\n");
    
    push(stack, 10);
    push(stack, 20);
    push(stack, 30);
    push(stack, 40);
    
    display(stack);
    
    printf("\nPeek: %d\n", peek(stack));
    
    printf("\nPopped: %d\n", pop(stack));
    printf("Popped: %d\n", pop(stack));
    
    display(stack);
    
    printf("\nComplexity Analysis:\n");
    printf("Push: O(1)\n");
    printf("Pop: O(1)\n");
    printf("Peek: O(1)\n");
    printf("Space: O(n)\n");
    
    free(stack);
    return 0;
}
